generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/JsonValidatorRequestFilter.java

```diff
@@ -0,0 +1,98 @@
+import java.io.IOException;
+import java.util.Optional;
+
+import javax.annotation.Priority;
+import javax.inject.Inject;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.Provider;
+
+import org.apache.commons.io.IOUtils;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+
+/**
+ * Filter that validates any JSON payload data in the request. Throws a BadRequestException if the payload
+ * has a JSON media type but fails to parse into valid JSON. Note validation does not involve JSON Schema,
+ * it is purely to check that the data is of a valid JSON format. This filter needs to run before
+ * LoggerRequestDataFilter as LoggerRequestDataFilter results in a HTTP 500 error code if it attempts to process invalid JSON
+ */
+@Priority(FILTER_PRIORITY)
+@Provider
+public class JsonValidatorRequestFilter implements ContainerRequestFilter {
+
+    public static final int FILTER_PRIORITY = USER - 100;
+
+    @Inject
+    Logger logger;
+
+    @Inject
+    HttpTraceLoggerHelper httpTraceLoggerHelper;
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
+        trace(logger, () -> "Validating JSON");
+        validateJsonPayloadIfPresent(requestContext);
+    }
+
+
+    private void validateJsonPayloadIfPresent(final ContainerRequestContext requestContext) throws IOException {
+        final Optional<MediaType> mediaType = Optional.ofNullable(requestContext.getMediaType());
+
+        if (isPayloadPresent(mediaType)) {
+            final String charset = mediaType.isPresent() ? charsetFrom(mediaType.get()) : defaultCharset().name();
+            final String payload = IOUtils.toString(requestContext.getEntityStream(), charset);
+
+            try {
+                if (isNotBlank(payload)) {
+                    new JSONObject(payload);
+                }
+            }
+            catch (JSONException jsonEx) {
+                final String message = format("Invalid JSON provided to [%s] JSON: [%s] ",
+                        httpTraceLoggerHelper.toHttpHeaderTrace(stripUserId(requestContext.getHeaders())),
+                        payload);
+                trace(logger, () -> message);
+                throw new BadRequestException(message);
+            }
+
+            requestContext.setEntityStream(new ByteArrayInputStream(payload.getBytes(charset)));
+        }
+
+    }
+
+    private boolean isPayloadPresent(final Optional<MediaType> mediaType) {
+        return mediaType.map(mediaTypeValue ->
+                null != mediaTypeValue.getSubtype() && mediaTypeValue.getSubtype().endsWith(JSON_MEDIA_TYPE_SUFFIX))
+                .orElse(false);
+    }
+
+    private static MultivaluedMap<String, String> stripUserId(final MultivaluedMap<String, String> headers) {
```

Comment:
Does this need to be static?
It could be pulled out as an class then injected as there seems to be two places where this code is repeated